src/pudl/analysis/timeseries_cleaning.py

```diff
@@ -1299,3 +1322,319 @@ def summarize_imputed(self, imputed: np.ndarray, mask: np.ndarray) -> pd.DataFra
                 }
             )
         return pd.DataFrame(stats)
+
+
+def flag_bad_values(df: pd.DataFrame) -> pd.DataFrame:
+    """Detect and null anomalous values in FERC 714 hourly demand matrix.
+
+    .. note::
+        Takes about 10 minutes.
+
+    Args:
+        df: FERC 714 hourly demand matrix,
+          as described in :func:`load_ferc714_hourly_demand_matrix`.
+
+    Returns:
+        Copy of `df` with nulled anomalous values.
+    """
+    ts = Timeseries(df)
+    ts.flag_ruggles()
+    return ts.to_dataframe(copy=False)
+
+
+def impute_flagged_values(df: pd.DataFrame, years: list[int]) -> pd.DataFrame:
+    """Impute null values in FERC 714 hourly demand matrix.
+
+    Imputation is performed separately for each year,
+    with only the respondents reporting data in that year.
+
+    .. note::
+        Takes about 15 minutes.
+
+    Args:
+        df: FERC 714 hourly demand matrix,
+          as described in :func:`load_ferc714_hourly_demand_matrix`.
+        years: list of years to input
+
+    Returns:
+        Copy of `df` with imputed values.
+    """
+    results = []
+    # sort here and then don't sort in the groupby so we can process
+    # the newer years of data first. This is so we can see early if
+    # new data causes any failures.
+    df = df.sort_index(ascending=False)
+    for year, gdf in df.groupby(df.index.year, sort=False):
+        # remove the records o/s of the working years because some
+        # respondents report one record of midnight of January first
+        # of the next year (report_date.dt.year + 1). and
+        # impute_ferc714_hourly_demand_matrix chunks over years at a time
+        # and having only one record
+        if year in years:
+            logger.info(f"Imputing year {year}")
+            keep = df.columns[~gdf.isnull().all()]
+            tsi = Timeseries(gdf[keep])
+            result = tsi.to_dataframe(tsi.impute(method="tnn"), copy=False)
+            results.append(result)
+    return pd.concat(results)
+
+
+def melt_imputed_matrix(df: pd.DataFrame) -> pd.DataFrame:
+    """Melt FERC 714 hourly demand matrix to long format.
+
+    Args:
+        df: FERC 714 hourly demand matrix,
+            as described in :func:`load_ferc714_hourly_demand_matrix`.
+        tz: FERC 714 respondent time zones,
+            as described in :func:`load_ferc714_hourly_demand_matrix`.
+
+    Returns:
+        Long-format hourly demand with columns ``respondent_id_ferc714``, report
+        ``year`` (int), ``datetime_utc``, and ``demand_mwh``.
```

Comment:
Need to update all the docstrings to be generic but you know that.

Reply:
I think I've fixed all the docstrings now